packages/service-library/tests/test_rabbitmq_robust_rpc.py

```diff
@@ -0,0 +1,229 @@
+        return CustomClass(x=self.x + other.x, y=self.y + other.y)
+
+
+@pytest.mark.parametrize(
+    "x,y,expected_result,expected_type",
+    [
+        pytest.param(12, 20, 32, int, id="expect_int"),
+        pytest.param(12, 20.0, 32.0, float, id="expect_float"),
+        pytest.param(b"123b", b"xyz0", b"123bxyz0", bytes, id="expect_bytes"),
+        pytest.param([1, 2], [2, 3], [1, 2, 2, 3], list, id="list_addition"),
+        pytest.param(
+            CustomClass(2, 1),
+            CustomClass(1, 2),
+            CustomClass(3, 3),
+            CustomClass,
+            id="custom_class",
+        ),
+        pytest.param(
+            CustomClass([{"p", "1"}], [{"h": 1}]),
+            CustomClass([{3, b"bytes"}], [{"b": 2}]),
+            CustomClass([{"p", "1"}, {3, b"bytes"}], [{"h": 1}, {"b": 2}]),
+            CustomClass,
+            id="custom_class_complex_objects",
+        ),
+    ],
+)
+async def test_base_rpc_pattern(
+    rpc_client: RobustRPCClient,
+    rpc_server: RobustRPCServer,
+    x: Any,
+    y: Any,
+    expected_result: Any,
+    expected_type: type,
+):
+    await rpc_server.register(add_me)
+
+    request_result = await rpc_client.request(add_me.__name__, x=x, y=y)
+    assert request_result == expected_result
+    assert type(request_result) == expected_type
+
+
+async def test_multiple_requests_sequence_same_server_and_client(
+    rpc_client: RobustRPCClient, rpc_server: RobustRPCServer
+):
+    await rpc_server.register(add_me)
+
+    for i in range(MULTIPLE_REQUESTS_COUNT):
+        assert await rpc_client.request(add_me.__name__, x=1 + i, y=2 + i) == 3 + i * 2
+
+
+async def test_multiple_requests_parallel_same_server_and_client(
+    rpc_client: RobustRPCClient, rpc_server: RobustRPCServer
+):
+    await rpc_server.register(add_me)
+
+    expected_result: list[int] = []
+    requests: list[Awaitable] = []
+    for i in range(MULTIPLE_REQUESTS_COUNT):
+        requests.append(rpc_client.request(add_me.__name__, x=1 + i, y=2 + i))
+        expected_result.append(3 + i * 2)
+
+    assert await asyncio.gather(*requests) == expected_result
+
+
+async def test_multiple_requests_parallel_same_server_different_clients(
+    rabbit_service: RabbitSettings, rpc_server: RobustRPCServer
+):
+    await rpc_server.register(add_me)
+
+    clients: list[RobustRPCClient] = []
+    for _ in range(MULTIPLE_REQUESTS_COUNT):
+        client = RobustRPCClient(rabbit_service)
+        clients.append(client)
+
+    # worst case scenario
+    await asyncio.gather(*[c.start() for c in clients])
+
+    requests: list[Awaitable] = []
+    expected_result: list[int] = []
+    for i in range(MULTIPLE_REQUESTS_COUNT):
+        client = clients[i]
+        requests.append(client.request(add_me.__name__, x=1 + i, y=2 + i))
+        expected_result.append(3 + i * 2)
+
+    assert await asyncio.gather(*requests) == expected_result
+
+    # worst case scenario
+    await asyncio.gather(*[c.stop() for c in clients])
```

Comment:
same here

Reply:
Also what about multiple servers? do they all get notified? or only one of them?

Reply:
> same here

same as above, they are ordered

Reply:
> Also what about multiple servers? do they all get notified? or only one of them?

I have a followup for multiple servers on the way